File: fons/codegen/py/norma/lista.ts
/**
 * Lista Method Registry - Python translations for Latin array methods
 *
 * COMPILER PHASE
 * ==============
 * codegen (Python target)
 *
 * ARCHITECTURE
 * ============
 * This module defines Python translations for lista<T> (array) methods.
 * Python uses list comprehensions and built-in functions rather than
 * method chaining for many operations.
 *
 * PYTHON IDIOMS
 * =============
 * | Latin         | Python                              |
 * |---------------|-------------------------------------|
 * | adde          | list.append(x)                      |
 * | addita        | [*list, x]                          |
 * | filtrata      | [x for x in list if pred(x)]        |
 * | mappata       | [f(x) for x in list]                |
 * | reducta       | functools.reduce(fn, list, init)    |
 * | primus        | list[0]                             |
 * | ultimus       | list[-1]                            |
 * | longitudo     | len(list)                           |
 * | continet      | x in list                           |
 * | inversa       | list[::-1]                          |
 * | ordinata      | sorted(list)                        |
 *
 * INPUT/OUTPUT CONTRACT
 * =====================
 * INPUT:  Latin method name from CallExpression
 * OUTPUT: Python code string
 * ERRORS: Returns undefined if method name not recognized
 */

// =============================================================================
// TYPES
// =============================================================================

export interface ListaMethod {
    latin: string;
    mutates: boolean;
    async: boolean;
    py: string | PyGenerator;
}

type PyGenerator = (obj: string, args: string) => string;

// =============================================================================
// METHOD REGISTRY
// =============================================================================

/**
 * Registry of Latin array methods with Python translations.
 */
export const LISTA_METHODS: Record<string, ListaMethod> = {
    // -------------------------------------------------------------------------
    // ADDING ELEMENTS
    // -------------------------------------------------------------------------

    adde: {
        latin: 'adde',
        mutates: true,
        async: false,
        py: 'append',
    },

    addita: {
        latin: 'addita',
        mutates: false,
        async: false,
        py: (obj, args) => `[*${obj}, ${args}]`,
    },

    praepone: {
        latin: 'praepone',
        mutates: true,
        async: false,
        py: (obj, args) => `${obj}.insert(0, ${args})`,
    },

    praeposita: {
        latin: 'praeposita',
        mutates: false,
        async: false,
        py: (obj, args) => `[${args}, *${obj}]`,
    },

    // -------------------------------------------------------------------------
    // REMOVING ELEMENTS
    // -------------------------------------------------------------------------

    remove: {
        latin: 'remove',
        mutates: true,
        async: false,
        py: 'pop',
    },

    remota: {
        latin: 'remota',
        mutates: false,
        async: false,
        py: obj => `${obj}[:-1]`,
    },

    decapita: {
        latin: 'decapita',
        mutates: true,
        async: false,
        py: obj => `${obj}.pop(0)`,
    },

    decapitata: {
        latin: 'decapitata',
        mutates: false,
        async: false,
        py: obj => `${obj}[1:]`,
    },

    purga: {
        latin: 'purga',
        mutates: true,
        async: false,
        py: 'clear',
    },

    // -------------------------------------------------------------------------
    // ACCESSING ELEMENTS
    // -------------------------------------------------------------------------

    primus: {
        latin: 'primus',
        mutates: false,
        async: false,
        py: obj => `${obj}[0]`,
    },

    ultimus: {
        latin: 'ultimus',
        mutates: false,
        async: false,
        py: obj => `${obj}[-1]`,
    },

    accipe: {
        latin: 'accipe',
        mutates: false,
        async: false,
        py: (obj, args) => `${obj}[${args}]`,
    },

    longitudo: {
        latin: 'longitudo',
        mutates: false,
        async: false,
        py: obj => `len(${obj})`,
    },

    vacua: {
        latin: 'vacua',
        mutates: false,
        async: false,
        py: obj => `len(${obj}) == 0`,
    },

    // -------------------------------------------------------------------------
    // SEARCHING
    // -------------------------------------------------------------------------

    continet: {
        latin: 'continet',
        mutates: false,
        async: false,
        py: (obj, args) => `(${args} in ${obj})`,
    },

    indiceDe: {
        latin: 'indiceDe',
        mutates: false,
        async: false,
        py: (obj, args) => `${obj}.index(${args})`,
    },

    inveni: {
        latin: 'inveni',
        mutates: false,
        async: false,
        py: (obj, args) => `next(filter(${args}, ${obj}), None)`,
    },

    inveniIndicem: {
        latin: 'inveniIndicem',
        mutates: false,
        async: false,
        py: (obj, args) => `next((i for i, x in enumerate(${obj}) if (${args})(x)), -1)`,
    },

    // -------------------------------------------------------------------------
    // TRANSFORMATIONS (return new list)
    // -------------------------------------------------------------------------

    filtrata: {
        latin: 'filtrata',
        mutates: false,
        async: false,
        py: (obj, args) => `list(filter(${args}, ${obj}))`,
    },

    mappata: {
        latin: 'mappata',
        mutates: false,
        async: false,
        py: (obj, args) => `list(map(${args}, ${obj}))`,
    },

    reducta: {
        latin: 'reducta',
        mutates: false,
        async: false,
        py: (obj, args) => {
            // Latin API: reducta(fn, init) - function first, then initial value
            // Split on LAST comma since init is typically simple (0, "", etc.)
            // while fn may contain commas in lambda params
            // NOTE: See README "Critical Note" - this is a workaround for the
            // architectural issue of receiving args as a joined string
            const lastComma = args.lastIndexOf(',');
            if (lastComma !== -1) {
                const fn = args.slice(0, lastComma).trim();
                const init = args.slice(lastComma + 1).trim();
                return `functools.reduce(${fn}, ${obj}, ${init})`;
            }
            // Just function, no initial value
            return `functools.reduce(${args}, ${obj})`;
        },
    },

    explanata: {
        latin: 'explanata',
        mutates: false,
        async: false,
        py: (obj, args) => `[y for x in ${obj} for y in (${args})(x)]`,
    },

    plana: {
        latin: 'plana',
        mutates: false,
        async: false,
        py: obj => `[y for x in ${obj} for y in x]`,
    },

    inversa: {
        latin: 'inversa',
        mutates: false,
        async: false,
        py: obj => `${obj}[::-1]`,
    },

    ordinata: {
        latin: 'ordinata',
        mutates: false,
        async: false,
        py: (obj, args) => {
            if (args) {
                return `sorted(${obj}, key=${args})`;
            }
            return `sorted(${obj})`;
        },
    },

    // -------------------------------------------------------------------------
    // SLICING
    // -------------------------------------------------------------------------

    sectio: {
        latin: 'sectio',
        mutates: false,
        async: false,
        py: (obj, args) => {
            // args is "start, end" or "start"
            const parts = args.split(',').map(s => s.trim());
            if (parts.length >= 2) {
                return `${obj}[${parts[0]}:${parts[1]}]`;
            }
            return `${obj}[${args}:]`;
        },
    },

    prima: {
        latin: 'prima',
        mutates: false,
        async: false,
        py: (obj, args) => `${obj}[:${args}]`,
    },

    ultima: {
        latin: 'ultima',
        mutates: false,
        async: false,
        py: (obj, args) => `${obj}[-${args}:]`,
    },

    omitte: {
        latin: 'omitte',
        mutates: false,
        async: false,
        py: (obj, args) => `${obj}[${args}:]`,
    },

    // -------------------------------------------------------------------------
    // PREDICATES
    // -------------------------------------------------------------------------

    omnes: {
        latin: 'omnes',
        mutates: false,
        async: false,
        py: (obj, args) => `all(map(${args}, ${obj}))`,
    },

    aliquis: {
        latin: 'aliquis',
        mutates: false,
        async: false,
        py: (obj, args) => `any(map(${args}, ${obj}))`,
    },

    // -------------------------------------------------------------------------
    // STRING OPERATIONS
    // -------------------------------------------------------------------------

    coniunge: {
        latin: 'coniunge',
        mutates: false,
        async: false,
        py: (obj, args) => {
            if (args) {
                return `${args}.join(${obj})`;
            }
            return `"".join(${obj})`;
        },
    },

    // -------------------------------------------------------------------------
    // ITERATION
    // -------------------------------------------------------------------------

    perambula: {
        latin: 'perambula',
        mutates: false,
        async: false,
        // forEach doesn't exist in Python; we use a for loop or list comprehension
        // For side effects, we use: [fn(x) for x in list] and discard result
        // Or generate a for loop if in statement context
        py: (obj, args) => `[(${args})(x) for x in ${obj}]`,
    },

    // -------------------------------------------------------------------------
    // IN-PLACE MUTATORS
    // -------------------------------------------------------------------------

    filtra: {
        latin: 'filtra',
        mutates: true,
        async: false,
        // Python list doesn't have in-place filter; use slice assignment
        py: (obj, args) => `${obj}[:] = [x for x in ${obj} if (${args})(x)]`,
    },

    ordina: {
        latin: 'ordina',
        mutates: true,
        async: false,
        py: (obj, args) => {
            if (args) {
                return `${obj}.sort(key=${args})`;
            }
            return `${obj}.sort()`;
        },
    },

    inverte: {
        latin: 'inverte',
        mutates: true,
        async: false,
        py: obj => `${obj}.reverse()`,
    },

    // -------------------------------------------------------------------------
    // ADVANCED OPERATIONS
    // -------------------------------------------------------------------------

    congrega: {
        latin: 'congrega',
        mutates: false,
        async: false,
        // groupBy - returns dict of lists
        py: (obj, args) => {
            // Python doesn't have native groupBy; use itertools or inline
            return `{k: list(g) for k, g in itertools.groupby(sorted(${obj}, key=${args}), key=${args})}`;
        },
    },

    unica: {
        latin: 'unica',
        mutates: false,
        async: false,
        py: obj => `list(dict.fromkeys(${obj}))`,
    },

    fragmenta: {
        latin: 'fragmenta',
        mutates: false,
        async: false,
        // chunk into sublists of size n
        py: (obj, args) => `[${obj}[i:i+${args}] for i in range(0, len(${obj}), ${args})]`,
    },

    densa: {
        latin: 'densa',
        mutates: false,
        async: false,
        // compact - remove falsy values
        py: obj => `[x for x in ${obj} if x]`,
    },

    partire: {
        latin: 'partire',
        mutates: false,
        async: false,
        // partition into [matching, non-matching]
        py: (obj, args) => `[[x for x in ${obj} if (${args})(x)], [x for x in ${obj} if not (${args})(x)]]`,
    },

    misce: {
        latin: 'misce',
        mutates: true,
        async: false,
        py: obj => `random.shuffle(${obj})`,
    },

    specimen: {
        latin: 'specimen',
        mutates: false,
        async: false,
        py: obj => `random.choice(${obj})`,
    },

    specimina: {
        latin: 'specimina',
        mutates: false,
        async: false,
        py: (obj, args) => `random.sample(${obj}, ${args})`,
    },

    // -------------------------------------------------------------------------
    // MATH OPERATIONS
    // -------------------------------------------------------------------------

    summa: {
        latin: 'summa',
        mutates: false,
        async: false,
        py: obj => `sum(${obj})`,
    },

    medium: {
        latin: 'medium',
        mutates: false,
        async: false,
        py: obj => `(sum(${obj}) / len(${obj}))`,
    },

    minimus: {
        latin: 'minimus',
        mutates: false,
        async: false,
        py: obj => `min(${obj})`,
    },

    maximus: {
        latin: 'maximus',
        mutates: false,
        async: false,
        py: obj => `max(${obj})`,
    },

    numera: {
        latin: 'numera',
        mutates: false,
        async: false,
        // count matching predicate
        py: (obj, args) => `sum(1 for x in ${obj} if (${args})(x))`,
    },
};

// =============================================================================
// LOOKUP FUNCTIONS
// =============================================================================

export function getListaMethod(name: string): ListaMethod | undefined {
    return LISTA_METHODS[name];
}

export function isListaMethod(name: string): boolean {
    return name in LISTA_METHODS;
}

File: subsidia/zig/lista.zig
// Lista - Dynamic array wrapper for Faber
// Wraps std.ArrayList with Latin method names
//
// WHY: No stored allocator. Methods that need allocation accept it as parameter.
// This ensures cura...fit blocks properly thread the current allocator.

const std = @import("std");

/// Lista(T) is a dynamic array with Latin-named methods.
/// Wraps std.ArrayList(T). Allocator passed per-method, not stored.
pub fn Lista(comptime T: type) type {
    return struct {
        items: std.ArrayList(T),

        const Self = @This();

        // =====================================================================
        // CONSTRUCTION
        // =====================================================================

        /// Create an empty lista.
        pub fn init(alloc: std.mem.Allocator) Self {
            return .{ .items = std.ArrayList(T).init(alloc) };
        }

        /// Create a lista from a slice.
        pub fn fromItems(alloc: std.mem.Allocator, data: []const T) Self {
            var self = init(alloc);
            self.items.appendSlice(data) catch @panic("OOM");
            return self;
        }

        /// Free the lista's memory.
        pub fn deinit(self: *Self) void {
            self.items.deinit();
        }

        /// Clone the lista (deep copy).
        pub fn clone(self: Self, alloc: std.mem.Allocator) Self {
            var result = init(alloc);
            result.items.appendSlice(self.items.items) catch @panic("OOM");
            return result;
        }

        // =====================================================================
        // ADDING ELEMENTS
        // =====================================================================

        /// Add element to end (mutates).
        pub fn adde(self: *Self, alloc: std.mem.Allocator, value: T) void {
            _ = alloc; // ArrayList tracks its allocator internally
            self.items.append(value) catch @panic("OOM");
        }

        /// Add element to end, returns new lista (immutable).
        pub fn addita(self: Self, alloc: std.mem.Allocator, value: T) Self {
            var result = self.clone(alloc);
            result.items.append(value) catch @panic("OOM");
            return result;
        }

        /// Add element to start (mutates).
        pub fn praepone(self: *Self, alloc: std.mem.Allocator, value: T) void {
            _ = alloc;
            self.items.insert(0, value) catch @panic("OOM");
        }

        /// Add element to start, returns new lista (immutable).
        pub fn praeposita(self: Self, alloc: std.mem.Allocator, value: T) Self {
            var result = self.clone(alloc);
            result.items.insert(0, value) catch @panic("OOM");
            return result;
        }

        // =====================================================================
        // REMOVING ELEMENTS
        // =====================================================================

        /// Remove and return last element (mutates).
        pub fn remove(self: *Self) ?T {
            return self.items.popOrNull();
        }

        /// Remove last element, returns new lista (immutable).
        pub fn remota(self: Self, alloc: std.mem.Allocator) Self {
            if (self.items.items.len == 0) return self.clone(alloc);
            var result = init(alloc);
            result.items.appendSlice(self.items.items[0 .. self.items.items.len - 1]) catch @panic("OOM");
            return result;
        }

        /// Remove and return first element (mutates).
        pub fn decapita(self: *Self) ?T {
            if (self.items.items.len == 0) return null;
            return self.items.orderedRemove(0);
        }

        /// Remove first element, returns new lista (immutable).
        pub fn decapitata(self: Self, alloc: std.mem.Allocator) Self {
            if (self.items.items.len == 0) return self.clone(alloc);
            var result = init(alloc);
            result.items.appendSlice(self.items.items[1..]) catch @panic("OOM");
            return result;
        }

        /// Clear all elements (mutates).
        pub fn purga(self: *Self) void {
            self.items.clearRetainingCapacity();
        }

        // =====================================================================
        // ACCESSING ELEMENTS
        // =====================================================================

        /// Get first element.
        pub fn primus(self: Self) ?T {
            if (self.items.items.len == 0) return null;
            return self.items.items[0];
        }

        /// Get last element.
        pub fn ultimus(self: Self) ?T {
            if (self.items.items.len == 0) return null;
            return self.items.items[self.items.items.len - 1];
        }

        /// Get element at index.
        pub fn accipe(self: Self, index: usize) ?T {
            if (index >= self.items.items.len) return null;
            return self.items.items[index];
        }

        /// Get raw slice for iteration.
        pub fn toSlice(self: Self) []T {
            return self.items.items;
        }

        // =====================================================================
        // PROPERTIES
        // =====================================================================

        /// Get length.
        pub fn longitudo(self: Self) usize {
            return self.items.items.len;
        }

        /// Check if empty.
        pub fn vacua(self: Self) bool {
            return self.items.items.len == 0;
        }

        // =====================================================================
        // SEARCHING
        // =====================================================================

        /// Check if contains element.
        pub fn continet(self: Self, value: T) bool {
            return std.mem.indexOfScalar(T, self.items.items, value) != null;
        }

        /// Find index of element.
        pub fn indiceDe(self: Self, value: T) ?usize {
            return std.mem.indexOfScalar(T, self.items.items, value);
        }

        // =====================================================================
        // PREDICATE METHODS
        // =====================================================================

        /// Check if all elements match predicate.
        pub fn omnes(self: Self, predicate: *const fn (T) bool) bool {
            for (self.items.items) |v| {
                if (!predicate(v)) return false;
            }
            return true;
        }

        /// Check if any element matches predicate.
        pub fn aliquis(self: Self, predicate: *const fn (T) bool) bool {
            for (self.items.items) |v| {
                if (predicate(v)) return true;
            }
            return false;
        }

        /// Find first element matching predicate.
        pub fn inveni(self: Self, predicate: *const fn (T) bool) ?T {
            for (self.items.items) |v| {
                if (predicate(v)) return v;
            }
            return null;
        }

        /// Find index of first element matching predicate.
        pub fn inveniIndicem(self: Self, predicate: *const fn (T) bool) ?usize {
            for (self.items.items, 0..) |v, i| {
                if (predicate(v)) return i;
            }
            return null;
        }

        // =====================================================================
        // FUNCTIONAL METHODS (allocating, return new lista)
        // =====================================================================

        /// Filter elements matching predicate.
        pub fn filtrata(self: Self, alloc: std.mem.Allocator, predicate: *const fn (T) bool) Self {
            var result = init(alloc);
            for (self.items.items) |v| {
                if (predicate(v)) result.items.append(v) catch @panic("OOM");
            }
            return result;
        }

        /// Map elements through transform function.
        /// Returns Lista of same type (T -> T transform).
        pub fn mappata(self: Self, alloc: std.mem.Allocator, transform: *const fn (T) T) Self {
            var result = init(alloc);
            for (self.items.items) |v| {
                result.items.append(transform(v)) catch @panic("OOM");
            }
            return result;
        }

        /// Reduce to single value.
        pub fn reducta(self: Self, reducer: *const fn (T, T) T, initial: T) T {
            var acc = initial;
            for (self.items.items) |v| {
                acc = reducer(acc, v);
            }
            return acc;
        }

        /// Reverse (returns new lista).
        pub fn inversa(self: Self, alloc: std.mem.Allocator) Self {
            var result = init(alloc);
            var i: usize = self.items.items.len;
            while (i > 0) {
                i -= 1;
                result.items.append(self.items.items[i]) catch @panic("OOM");
            }
            return result;
        }

        /// Sort (returns new lista).
        pub fn ordinata(self: Self, alloc: std.mem.Allocator) Self {
            const result = self.clone(alloc);
            std.mem.sort(T, result.items.items, {}, std.sort.asc(T));
            return result;
        }

        /// Sort with custom comparator (returns new lista).
        pub fn ordinataCum(self: Self, alloc: std.mem.Allocator, comptime lessThan: fn (void, T, T) bool) Self {
            const result = self.clone(alloc);
            std.mem.sort(T, result.items.items, {}, lessThan);
            return result;
        }

        /// Slice - take elements from start to end.
        pub fn sectio(self: Self, alloc: std.mem.Allocator, start: usize, end: usize) Self {
            var result = init(alloc);
            const s = @min(start, self.items.items.len);
            const e = @min(end, self.items.items.len);
            if (s < e) {
                result.items.appendSlice(self.items.items[s..e]) catch @panic("OOM");
            }
            return result;
        }

        /// Take first n elements.
        pub fn prima(self: Self, alloc: std.mem.Allocator, n: usize) Self {
            const count = @min(n, self.items.items.len);
            var result = init(alloc);
            result.items.appendSlice(self.items.items[0..count]) catch @panic("OOM");
            return result;
        }

        /// Take last n elements.
        pub fn ultima(self: Self, alloc: std.mem.Allocator, n: usize) Self {
            const count = @min(n, self.items.items.len);
            const start = self.items.items.len - count;
            var result = init(alloc);
            result.items.appendSlice(self.items.items[start..]) catch @panic("OOM");
            return result;
        }

        /// Skip first n elements.
        pub fn omitte(self: Self, alloc: std.mem.Allocator, n: usize) Self {
            const skip = @min(n, self.items.items.len);
            var result = init(alloc);
            result.items.appendSlice(self.items.items[skip..]) catch @panic("OOM");
            return result;
        }

        // =====================================================================
        // MUTATING OPERATIONS
        // =====================================================================

        /// Sort in place.
        pub fn ordina(self: *Self) void {
            std.mem.sort(T, self.items.items, {}, std.sort.asc(T));
        }

        /// Sort in place with custom comparator.
        pub fn ordinaCum(self: *Self, comptime lessThan: fn (void, T, T) bool) void {
            std.mem.sort(T, self.items.items, {}, lessThan);
        }

        /// Reverse in place.
        pub fn inverte(self: *Self) void {
            std.mem.reverse(T, self.items.items);
        }

        // =====================================================================
        // ITERATION
        // =====================================================================

        /// Execute callback for each element.
        pub fn perambula(self: Self, callback: *const fn (T) void) void {
            for (self.items.items) |v| {
                callback(v);
            }
        }

        // =====================================================================
        // AGGREGATION (numeric types only - caller must ensure T is numeric)
        // =====================================================================

        /// Sum of numeric elements.
        /// Only valid for integer types.
        pub fn summa(self: Self) T {
            var sum: T = 0;
            for (self.items.items) |v| {
                sum += v;
            }
            return sum;
        }

        /// Average of numeric elements.
        /// Returns f64 for precision.
        pub fn medium(self: Self) f64 {
            if (self.items.items.len == 0) return 0.0;
            var sum: T = 0;
            for (self.items.items) |v| {
                sum += v;
            }
            return @as(f64, @floatFromInt(sum)) / @as(f64, @floatFromInt(self.items.items.len));
        }

        /// Minimum value.
        pub fn minimus(self: Self) ?T {
            return std.mem.min(T, self.items.items);
        }

        /// Maximum value.
        pub fn maximus(self: Self) ?T {
            return std.mem.max(T, self.items.items);
        }

        /// Count elements matching predicate.
        pub fn numera(self: Self, predicate: ?*const fn (T) bool) usize {
            if (predicate) |pred| {
                var count: usize = 0;
                for (self.items.items) |v| {
                    if (pred(v)) count += 1;
                }
                return count;
            }
            return self.items.items.len;
        }
    };
}
